cmd/maya-apiserver/cstor-operator/controller.go

```diff
@@ -71,112 +67,121 @@ type Controller struct {
+	return cb
+}
+
+// withWorkqueue adds workqueue to controller object.
+func (cb *ControllerBuilder) withWorkqueue() *ControllerBuilder {
+	cb.Controller.workqueue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "SPC")
+	return cb
+}
+
+// withRecorder adds recorder to controller object.
+func (cb *ControllerBuilder) withRecorder(ks kubernetes.Interface) *ControllerBuilder {
 	openebsScheme.AddToScheme(scheme.Scheme)
 	glog.V(4).Info("Creating event broadcaster")
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(glog.Infof)
-	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: ks.CoreV1().Events("")})
 	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
-	queueLoad := QueueLoad{}
-	controller := &Controller{
-		kubeclientset: kubeclientset,
-		clientset:     clientset,
-		deletedIndexer: cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc,
-			cache.Indexers{}),
-		spcSynced: spcInformer.Informer().HasSynced,
-		workqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "SPC"),
-		recorder:  recorder,
-		queueLoad: queueLoad,
-	}
-
-	glog.Info("Setting up event handlers")
+	cb.Controller.recorder = recorder
+	return cb
+}
 
+// withEventHandler adds event handlers controller object.
+func (cb *ControllerBuilder) withEventHandler(spcInformerFactory informers.SharedInformerFactory) *ControllerBuilder {
+	spcInformer := spcInformerFactory.Openebs().V1alpha1().StoragePoolClaims()
 	// Set up an event handler for when SPC resources change
 	spcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: controller.addSpc,
-
-		// Informer will send update event along with object in following cases:
-		// 1. In case the object is updated ( Change of Resource Version)
-
-		// 2. In case the object is deleted by using a finalizer.
-		//    Some of the kubectl version e.g. v1.11.0 will make use of a finalizer beforehand deletion.
-		//    Hence in this case the delete of an object will cause a invoke of UpdateFunc
-		//    as the finalizer addition will cause Resource Version to change.
-		//
-		//    But some of the kubectl version will not make use of finalizer for
-		//    object deletion.
-		//    Whatever be the case, delete of the resource will always invoke DeleteFunc
-		//    and hence the delete event will be captured by it and the delete event
-		//    that will be delivered in UpdateFunc due to specific kubectl versions will
-		//    be suppressed.(see in updateSpc function, where delete event is marked as ignoreEvent)
-
-		// 3. After every fixed amount of time which is know as reSync Period.
-		//    ReSync period can be set to values we want. It can help in reconiciliation.
-		UpdateFunc: controller.updateSpc,
```

Comment:
Due to removal of finalizers, does v0.8.x branch have issue as there is no callback for deletefunc?

Reply:
@AmitKumarDas ^^

Reply:
Finalizers were never present.